### ISSUE TITLE
ibdata1 growing too fast

### ISSUE THREAD

**Issue by kashizui** [codalab/codalab-worksheets#573]
https://www.percona.com/blog/2013/08/20/why-is-the-ibdata1-file-continuously-growing-in-mysql/

I have a suspicion that worker check-ins or something are creating never-closed transactions, etc. (maybe related to #571)

**Comment by kashizui** [comment]
ibdata1 file has grown from zero to 7.3gb on wstest in the last two weeks, this is probably an unreasonable amount? I'm not actually very familiar with InnoDB, this requires more research.

**Comment by kashizui** [comment]
Running the innochecksum program provided here 
http://bugs.mysql.com/bug.php?id=57611

I get this on wsprod:

```
codalabwsprod:~% sudo ./innochecksum /codalabdata/mysql/ibdata1
bytes read (8192) doesn't match universal page size (16384)
0   bad checksum
36121   FIL_PAGE_INDEX
9727    FIL_PAGE_UNDO_LOG
10  FIL_PAGE_INODE
38  FIL_PAGE_IBUF_FREE_LIST
202241  FIL_PAGE_TYPE_ALLOCATED
98  FIL_PAGE_IBUF_BITMAP
137 FIL_PAGE_TYPE_SYS
2   FIL_PAGE_TYPE_TRX_SYS
3   FIL_PAGE_TYPE_FSP_HDR
102 FIL_PAGE_TYPE_XDES
1338485 FIL_PAGE_TYPE_BLOB
0   FIL_PAGE_TYPE_ZBLOB
0   other
219 max index_id
undo type: 786 insert, 8941 update, 0 other
undo state: 0 active, 150 cached, 337 to_free, 1561 to_purge, 0 prepared, 7679 other
#pages  index_id
1   1
1   2
1   3
1   4
1   5
1   11
1   12
1   13
1   14
1   1562
83  1563
17  1564
18  1565
61  1568
28  1569
418 1570
321 1571
152 1572
21367   1573
392 1574
1246    1575
256 1576
601 1577
425 1578
1   1579
1   1580
1   1581
1   1582
30  1583
15  1584
18  1585
5   1586
3   1587
3   1588
5   1592
3   1593
1   1594
1   1595
149 1596
11  1597
72  1598
69  1599
1   1600
1   1601
1   1602
14  1700
1   2258
1   2259
1   2260
1   2261
1   2262
1   2264
1   2265
1   2266
1   2269
1   2270
1   2272
3   2273
1   2274
1   2275
1   2276
1   2277
1   2278
1   2279
1   2280
33  2305
5   2306
5   2307
5   2308
1   2309
1   2310
1   2311
1   2312
1   2313
11  2315
6   2316
1095    2317
214 2318
2   2321
6   2322
7819    2323
102 2324
396 2325
95  2326
178 2327
103 2328
```

Whatever `FIL_PAGE_TYPE_BLOB` is, it's also taking up most of the tablespace on wstest.
Pending further investigation

**Comment by kashizui** [comment]
Converted a lot of `TEXT` columns on wstest to `TINYTEXT`, we'll see how ibdata1 grows from now on. It is 11gb as of today.

**Comment by percyliang** [comment]
Thanks for looking into this.  Can we do a controlled experiment where we
hit the DB hard to make it grow with TEXT, and see what happens with
TINYTEXT with the same load?

**Comment by kashizui** [comment]
Reducing the size of the `TEXT` fields will not do much, because these variable-length fields do not preallocate space -- they work similarly to malloc and only use as much space as its contents.

I think this has to do with tablespace fragmentation due to frequent deletes/updates of rows with variable length columns (e.g. TEXT), which has been occurring a lot due to the new worker system. This is further evidenced by the fact that `FIL_PAGE_TYRPE_BLOB` dominates the disk usage in `ibdata1`.

#607 has a few changes to minimize deletes of worker-related rows. However we will still be frequently updating the TEXT/BLOB columns.

A better solution is to use the InnoDB File-Per-Table Tablespaces feature, and periodically defragment the affected tables by running `OPTIMIZE TABLE` (this does not do very much now because all of our tables are in a single file `ibdata1`). This is actually now the default setup from MySQL version 5.6.6 and onwards.
https://dev.mysql.com/doc/refman/5.6/en/innodb-multiple-tablespaces.html

If that sounds good I'll go ahead and make these changes as soon as possible, which will help us reclaim our disk space as well.

**Comment by kashizui** [comment]
I found the real reason for this and it doesn't have very much to do with tablespace fragmentation. I can't believe I missed it before. The huge JSON blobs that the workers send on checkin are being saved every time into the event table verbatim in the args column. The args column is 64K per record, and workers checkin every half second, which means that with 6 workers, the event table is growing by a minimum of 46 MB per hour, which makes more than 1 GB per day just from checkin events.

This query, measuring the size of the events table on wsprod, confirms this:

```
>> SELECT (data_length+index_length)/power(1024,3) tablesize_gb
FROM information_schema.tables
WHERE table_schema='codalab_bundles' and table_name='event';

165.44012451171875
```

(The entire MySQL database is currently 168gb.)

Potential steps to recover and fix:
- [ ] Truncate the event table, then redump and reload the tables to recover some disk space
- [ ] Implement some diffing/compression procedure to fix #608 
- [ ] Implement special logging policies for commands like checkin (e.g. severely truncating the content of the args column, throttling the logging frequency)

I think completely blacklisting the checkin command from the event log is not a good idea, since it was super valuable to be able to query the event table for some statistics on runtime, especially when implementing #607 .

@percyliang  thoughts?

**Comment by kashizui** [comment]
Final Steps:

1. Deleted all checkin commands from the event table using this iterative query:
```
delimiter $$
DROP PROCEDURE IF EXISTS DeleteCheckins $$
CREATE PROCEDURE DeleteCheckins()
BEGIN
SET @deleted_total = 0;
REPEAT
	DELETE FROM `event`
	WHERE command = "POST /workers/<worker_id>/checkin"
	ORDER BY `id`
	LIMIT 1000;
	SET @deleted = ROW_COUNT();
	SET @deleted_total = @deleted_total + @deleted;
	SELECT @deleted_total;
UNTIL @deleted = 0 END REPEAT;
END $$
CALL DeleteCheckins $$
```
2.  Rebuilt the database using https://github.com/codalab/codalab-worksheets/pull/232

Result: Reclaimed over 160gb of diskspace!
